src/Stream.php

```diff
@@ -260,11 +260,19 @@ public function getContents(): string
             throw new \RuntimeException('Stream is detached');
         }
 
-        if (false === $contents = @\stream_get_contents($this->stream)) {
-            throw new \RuntimeException('Unable to read stream contents: ' . (\error_get_last()['message'] ?? ''));
+        $exception = null;
+
+        \set_error_handler(static function ($type, $message) use (&$exception) {
+            throw $exception = new \RuntimeException('Unable to read stream contents: ' . $message);
+        });
+
+        try {
+            return \stream_get_contents($this->stream);
+        } catch (\Throwable $e) {
+            throw $e === $exception ? $e : new \RuntimeException('Unable to read stream contents: ' . $e->getMessage(), 0, $e);
+        } finally {
+            \restore_error_handler();
```

Comment:
I don't understand how setting an error handler fixes this issue when [other tools](https://github.com/thecodingmachine/safe/blob/3a6e43b4be3d3d65b272c2660300a50658ce168b/generated/stream.php#L221-L235) and [even Guzzle](https://github.com/guzzle/psr7/blob/2.6/src/Utils.php#L420-L422) check for a `false` return. Nothing about this code seems like it will actually solve the issue, because the return value of `stream_get_contents` is not being checked.

Reply:
When stream_get_contents returns false, it first always trigger a PHP warning. This we catch those in the error handler, the function will never really return false.

Reply:
Considering that such behavior is [not documented](https://www.php.net/stream_get_contents), why take that approach over checking the return value is false? Every other fix I've looked at around this issue is doing a `$ret === false` check...

Reply:
> That would be dead code.

Are you saying that checking `=== false` does not work, or are you saying that this monkey patching with the error handler identifies the _impending_ `return false` more quickly?

Reply:
I'm saying that the function will never return false because the error handler will always throw before.

Reply:
But, if the error handler monkey patching was not there, we could rely on `$ret === false`, couldn't we?

In any case, I feel like either:

- this code needs a comment describing what it does, given that it relies on undocumented behavior
- it should rely on `$ret === false` checks, which the code did prior to this patch (but perhaps not correctly?)

Reply:
Also, I want to be clear that I really appreciate you identifying this issue and submitting the PR. It only bothers me that we have MANY disparate projects with different solutions to the same problem... perhaps we should all be using [safe](https://github.com/thecodingmachine/safe) instead?

Reply:
If we don't set our own error handler, then we rely the existing, which can be a custom and broken one. But putting our own for the operation, we make things safe.

Reply:
See the link to the PHP code in the PR description. When doing a partial read, PHP does not return false right now because it does not propagate the `false` return value.

Reply:
@stof 😕 and is there a related PHP bug report and/or PR?

Reply:
> perhaps we should all be using (and contributing to) [safe](https://github.com/thecodingmachine/safe) instead?

"safe" is too opinionated to be used in a generic package IMHO ;)

Reply:
@nicolas-grekas the current "use this arcane code instead" situation being applied to a multitude of projects doesn't seem like a good solution either. Would adding `safe` really be such a terrible thing? And if safe is too opinionated, why not contribute your knowledge to making it something that package developers would support?

Reply:
Considering `safe` is completely opt-in (you have to explicitly use the `Safe\\` namespace to use the functions it guards), it seems like it's up to the individual package maintainers as to whether or not it's "too opinionated".

Reply:
I have a question to understand this patch a little better. 
Why do you declare an "$exception" variable for the Exception although the Catch block gets the same Exception instance? Is there some "PHPish" specific logic / behavior behind this (that is not documented)?

Reply:
If we remove this check, we would always wrap the exception thrown in the error handler in another exception. This check allows removing this needless wrapping exception.

Reply:
Ok, thanks. Then my next question is, why not move the `set_error_handler` into the try-block block to handle that scope in the same place?

Reply:
if set_error_handler fails for any reasons, we don't want restore_error_handler to be called

Reply:
And in PHP, try/catch does not create a scope anyway.